cmd/mtaresolver.go

```diff
@@ -0,0 +1,392 @@
+	// if source.Type == RESOURCE_TYPE && source.Resource.Type == "configuration" {
+	// 	missingType = "Unresolved-conf/" + source.Resource.Name
+	// }
+
+	// return "~{" + missingType + ": " + variableName + "}"
+}
+
+func (m *MTAResolver) resolvePlaceholders(sourceModule *mta.Module, source *MTASource, requires *mta.Requires, valueObj interface{}) interface{} {
+	switch valueObj.(type) {
+	case map[interface{}]interface{}:
+		v := ConvertToJsonSafe(valueObj)
+		return m.resolvePlaceholders(sourceModule, source, requires, v)
+	case map[string]interface{}:
+		value := valueObj.(map[string]interface{})
+		for k, v := range value {
+			value[k] = m.resolvePlaceholders(sourceModule, source, requires, v)
+		}
+		return value
+	case []interface{}:
+		value := valueObj.([]interface{})
+		for k, v := range value {
+			value[k] = m.resolvePlaceholders(sourceModule, source, requires, v)
+		}
+		return value
+	case string:
+		return m.resolvePlaceholdersString(sourceModule, source, requires, valueObj.(string))
+	default:
+		//if the value is not a string but a leaf, just return it
+		return valueObj
+	}
+}
+
+func (m *MTAResolver) resolvePlaceholdersString(sourceModule *mta.Module, source *MTASource, requires *mta.Requires, value string) interface{} {
+	pos := 0
+	pos, placeholderName, wholeValue := ParseNextVariable(pos, value, PLACEHOLDER_PREFIX)
+
+	if pos < 0 {
+		return value
+	}
+	placeholderValue := m.getParameter(sourceModule, source, requires, placeholderName)
+
+	if wholeValue {
+		return placeholderValue
+	}
+	for pos >= 0 {
+		phValueStr, _ := convertToString(placeholderValue)
+		value = value[:pos] + phValueStr + value[pos+len(placeholderName)+3:]
+		//todo verify test coverage
+		pos, placeholderName, wholeValue = ParseNextVariable(pos+len(phValueStr), value, PLACEHOLDER_PREFIX)
+		if pos >= 0 {
+			placeholderValue = m.getParameter(sourceModule, source, requires, placeholderName)
+		}
+	}
+
+	return value
+}
+
+func (m *MTAResolver) getParameter(sourceModule *mta.Module, source *MTASource, requires *mta.Requires, paramName string) string {
+	//first on source parameters scope
+	if source != nil {
+		paramVal := source.Parameters[paramName]
+		if paramVal != nil {
+			return paramVal.(string)
+		}
+
+		//defaults to context's module params:
+		paramValStr, ok := m.context.modules[source.Name][paramName]
+		if ok {
+			return paramValStr
+		}
+
+		//defaults to context's resource params:
+		paramValStr, ok = m.context.resources[source.Name][paramName]
+		if ok {
+			return paramValStr
+		}
+	}
+
+	//then try on requires level
+	if requires != nil {
+		paramVal := requires.Parameters[paramName]
+		if paramVal != nil {
+			return paramVal.(string)
+		}
+	}
+
+	if sourceModule != nil {
+		paramVal := sourceModule.Parameters[paramName]
+		if paramVal != nil {
+			return paramVal.(string)
+		}
+		//defaults to context's module params:
+		paramValStr, ok := m.context.modules[sourceModule.Name][paramName]
+		if ok {
+			return paramValStr
+		}
```

Comment:
isn't search on m.context.resources relevant?

Reply:
No. within a module, you search in the module, or in its required resource or module, or globally. but not arbitrary resource which is not required...